File: sections/53_advanced_combinatorics/problems/d_unlabeled_teams/solution.py
import sys


MOD = 1_000_000_007


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n, k = data
    dp = [0] * (k + 1)
    dp[0] = 1
    for people in range(1, n + 1):
        for teams in range(min(people, k), 0, -1):
            dp[teams] = (dp[teams - 1] + teams * dp[teams]) % MOD
        dp[0] = 0
    print(dp[k])


if __name__ == "__main__":
    main()

File: sections/53_advanced_combinatorics/problems/d_unlabeled_teams/solution.cpp
#include <bits/stdc++.h>
using namespace std;

const long long MOD = 1'000'000'007LL;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, k;
    if (!(cin >> n >> k)) return 0;
    vector<long long> dp(k + 1, 0);
    dp[0] = 1;
    for (int people = 1; people <= n; people++) {
        for (int teams = min(people, k); teams >= 1; teams--) {
            dp[teams] = (dp[teams - 1] + teams * dp[teams]) % MOD;
        }
        dp[0] = 0;
    }
    cout << dp[k] << '\n';
    return 0;
}
